Automattic/newspack-network
# Feat/network pass

Creates the Network pass feature.

Also, this work surfaced a small issue with the plugin. The post_process_in_hub method was being called even for events that were triggered in the Hub itself. This would cause the local memberhips to get the post meta and they would become uneditable.

To be consistent, this method should not be called in such cases. Just as process_in_node is not called in node for their own events.

I slightly changed the architecture there and updated the docs. It should be good now.

** Testing

Setup

* On site A, create two membership plans
* In one of the plans, add any value to the new "Network ID" field

![image](https://github.com/Automattic/newspack-network/assets/971483/db54a04f-5be3-4c6a-9566-d9dbe7908f0e)


* Leave Network ID blank for the other membership plan
* Do the same on Site B, two memberships
* In one of the memberships, put the same Network ID

Now let's test the flow

* Create a new membership for a user in one of the sites - in the memberhips with a Network ID
* Confirm that a `newspack_network_woo_membership_updated` is triggered and added to the event log
* Once this event reaches the other site, confirm the user is created and a membership is granted to them in the membership plan that shares the same network id
* In the site where the membership was distributed to, confirm the link to edit the membership points to the original site, where the membership was first created
* Confirm the action links (when you hover the memberhip in the admin table) shows a message "Managed in {link}" with a link to the original membership

![image](https://github.com/Automattic/newspack-network/assets/971483/4c3131f1-6c1d-4f01-887d-f8bf97821999)

* Edit the orginal membership and change its status and confirm all status changes are properly propagated across the network
* Create a membership in the membership plan without a Network ID and confirm no events are triggered
